rotkehlchen/chain/evm/decoding/odos/v2/decoder.py

```diff
@@ -0,0 +1,242 @@
+            msg_aggregator=msg_aggregator,
+        )
+        self.evm_txns = EvmTransactions(self.evm_inquirer, self.base.database)
+        self.router_address = router_address
+        self.native_currency = self.evm_inquirer.native_token
+
+    def _resolve_tokens_data(
+            self,
+            token_addresses: list['ChecksumEvmAddress'],
+            token_amounts: list[int],
+    ) -> dict[str, FVal]:
+        """Returns the resolved evm tokens or native currency and their amounts"""
+        resolved_result: dict[str, FVal] = {}
+        for token_address, token_amount in zip(token_addresses, token_amounts, strict=True):
+            if token_address == ZERO_ADDRESS:
+                asset = self.native_currency
+            else:
+                asset = get_or_create_evm_token(
+                    userdb=self.base.database,
+                    evm_inquirer=self.base.evm_inquirer,
+                    evm_address=token_address,
+                    chain_id=self.base.evm_inquirer.chain_id,
+                    token_kind=EvmTokenKind.ERC20,
+                )
+
+            resolved_result[asset.identifier] = asset_normalized_value(amount=token_amount, asset=asset)  # noqa: E501
+
+        return resolved_result
+
+    def _calculate_router_fee(
+            self,
+            context: 'DecoderContext',
+            sender: 'ChecksumEvmAddress',
+            output_tokens: dict[str, FVal],
+    ) -> list[EvmEvent]:
+        """Calculates the fees collected by the router and returns the fee events.
+        Fees are the amount of out tokens that router received - the amount that router sent.
+        `output_tokens` is a dictionary mapping from token identifier to the amount of tokens
+        that were received by the user.
+        """
+        router_holding_amount: defaultdict[Asset, FVal] = defaultdict(FVal)
+        for tx_log in context.all_logs:
+            if (
+                tx_log.topics[0] != ERC20_OR_ERC721_TRANSFER or
+                ((identifier := evm_address_to_identifier(
+                    address=tx_log.address,
+                    chain_id=self.evm_inquirer.chain_id,
+                    token_type=EvmTokenKind.ERC20,
+                )) not in output_tokens)
+            ):
+                continue
+
+            # can directly initialize EvmToken here because it's in output_tokens which were get_or_created above  # noqa: E501
+            asset = EvmToken(identifier)
+            if hex_or_bytes_to_address(tx_log.topics[2]) == self.router_address:
+                router_holding_amount[asset] += asset_normalized_value(amount=hex_or_bytes_to_int(tx_log.data), asset=asset)  # noqa: E501
+
+            elif hex_or_bytes_to_address(tx_log.topics[1]) == self.router_address:
+                router_holding_amount[asset] -= asset_normalized_value(amount=hex_or_bytes_to_int(tx_log.data), asset=asset)  # noqa: E501
+
+        if self.native_currency.identifier in output_tokens:
+            try:
```

Comment:
Perhaps add a small comment
```suggestion
            try:  # download (if not there) and find all native token transfers from/to the router
```